### ISSUE TITLE
Automated learning algorithm overview

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/sktime#704]
Contributors: @fkiraly, @mloning, @afzal442, @abdulelahsm, @sparkingdark 

## idea:
web data base for learning algorithms with fields for: 
* class name, 
* scitype (e.g. forecaster, classifier, etc.), 
* original contributor,
* maintainer/current [code owner](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners),
* link to API reference,

optional additional fields: 
* links to literature references, 
* package dependency, 
* code health/status (under development, mature, etc), 
* link to example notebooks

## use cases:
* search for suitable algorithms given a problem
* search for information about a given algorithm

## reference example from mlr3 library:
* https://mlr3extralearners.mlr-org.com/articles/learners/learner_status.html
* https://mlr3extralearners.mlr-org.com/articles/learners/list_learners.html
* GitHub action to automatically generate table: https://github.com/mlr-org/mlr3extralearners/blob/main/.github/workflows/update_learner_table.yml

## related code in sktime:
```python
from sktime.utils import all_estimators
all_estimators(estimator_types="classifier")
```
https://github.com/alan-turing-institute/sktime/blob/ee7a06843a44f4aaec7582d847e36073a9ab0566/sktime/utils/__init__.py#L16
## questions:
* integrate it with API reference: https://www.sktime.org/en/latest/api_reference.html (adding columns to that table)?

## notes:
* make clear that not all algorithm are reference implementations, even though we link them with the original research paper, some may be re-implemented, adapted or interfaced from other packages
* encourage people to ping algorithm maintainers directly when they have questions about algorithm using the GitHub user name
* Adding col for maintainer using GH

**Comment by mloning** [comment]
Related issue: https://github.com/alan-turing-institute/sktime/issues/354

**Comment by fkiraly** [comment]
@mloning, would you mind assigning to whoever is currently looking at this?

**Comment by mloning** [comment]
@fkiraly still the same people, @afzal442, Abdul and @sparkingdark

**Comment by fkiraly** [comment]
I mean, assigning to the issue

**Comment by mloning** [comment]
@fkiraly I think they need to comment on the issue first before they can be assigned for some reason.

**Comment by afzal442** [comment]
Thanks @fkiraly.
Hi @mloning , since we have been working on this issue, you can assign it to us as @abdulelahsm, @afzal442 , @sparkingdark. Thanks.

**Comment by fkiraly** [comment]
hm, it would appear @sparkingdark and @abdulelahsm have to post sth here in this thread to be assignable.

**Comment by mloning** [comment]
The related OLS issue is here: https://github.com/open-life-science/ols-3/issues/36

**Comment by abdulelahsm** [comment]
.

**Comment by afzal442** [comment]
Hi @mloning, we have come up with some approaches towards our project that we aim to sum up in recent weeks. We both (Abdul and I) with our mentor have been able to discuss our ideas so far and apply the same with regards to OLS. 

```
from sktime.utils import all_estimators
from importlib import import_module

for modname, modclass in all_estimators():
    try:
        print(modname, modclass.__author__)
    except AttributeError:
        try:
            print(modname, import_module(modclass.__module__).__author__)
        except AttributeError:
            print(modname, "no author info")
```
**Output:**
```
ARIMA ['Markus Löning', 'Hongyi Yang']
AutoARIMA ['Markus Löning', 'Hongyi Yang']
AutoCorrelationTransformer ['Afzal Ansari']
AutoETS ['Hongyi Yang']
BATS ['Martin Walter']
........
```

```
from sktime.utils import all_estimators
from importlib import import_module

estimators = all_estimators(estimator_types="classifier")
# Depends on types of estimator

for modname, modclass in estimators:
    try:
        print(modname, modclass, modclass.__author__)
    except AttributeError:
        try:
            print(modname, modclass, import_module(modclass.__module__).__author__)
        except AttributeError:
            print(modname, modclass, "no author info")
```
**Output:**
```
BOSSEnsemble <class 'sktime.classification.dictionary_based._boss.BOSSEnsemble'> Matthew Middlehurst
ColumnEnsembleClassifier <class 'sktime.classification.compose._column_ensemble.ColumnEnsembleClassifier'> ['Aaron Bostrom']
ComposableTimeSeriesForestClassifier <class 'sktime.classification.compose._ensemble.ComposableTimeSeriesForestClassifier'> ['Markus Löning', 'Ayushmaan Seth'].........
```.

**Comment by afzal442** [comment]
**Writing into CSV file to generate table further:**

```
   ...: from sktime.utils import all_estimators 
   ...: from importlib import import_module 
   ...: import csv 
   ...:  
   ...: with open("author_list.csv", "w+", newline='') as csvfile: 
   ...:     write = csv.writer(csvfile, delimiter=",")  
   ...:     write.writerow(["Name","Class_ID", "Authors"])  
   ...:     for modname, modclass in all_estimators(): 
   ...:         try: 
   ...:             write = csv.writer(csvfile, delimiter = ",") 
   ...:             write.writerow([modname,modclass, modclass.__author__]) 
   ...:         except AttributeError: 
   ...:             try:                      
   ...:                 write = csv.writer(csvfile, delimiter = ",") 
   ...:                 write.writerow([modname,modclass, str(import_module(modclass.__module__).__author__)]) 
   ...:                  
   ...:             except AttributeError: 
   ...:                 write = csv.writer(csvfile, delimiter = ",")
   ...:                 write.writerow([modname,modclass, "no author info"])
```
O/P:
```
| Name                                      | Class_ID                                                                                             | Authors                                                                               |
| ----------------------------------------- | ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- |
| ARIMA                                     | <class 'sktime.forecasting.arima.ARIMA'>                                                             | ['Markus Löning', 'Hongyi Yang']                                                      |
| AutoARIMA                                 | <class 'sktime.forecasting.arima.AutoARIMA'>                                                         | ['Markus Löning', 'Hongyi Yang']                                                      |
| AutoCorrelationTransformer                | <class 'sktime.transformations.series.acf.AutoCorrelationTransformer'>                               | ['Afzal Ansari']                                                                      |
| AutoETS                                   | <class 'sktime.forecasting.ets.AutoETS'>                                                             | ['Hongyi Yang']                                                                       |
| BATS                                      | <class 'sktime.forecasting.bats.BATS'>                                                               | ['Martin Walter']                                                                     |
| BOSSEnsemble                              | <class 'sktime.classification.dictionary_based._boss.BOSSEnsemble'>                                  | Matthew Middlehurst                                                                   |
```

**Comment by tobyhodges** [comment]
It looks like the format of the `__author__` fields are not standardised across the different estimators in `sktime`: some are lists of multiple names (e.g. lines 1&2 above), some are lists with only one name (e.g. lines 3-5 above) and some are strings with a single name (e.g. line 6 above). Furthermore, we noticed that a few individual estimator functions have their own `__author__` field, but that it is more common for the `__author__` to be defined at the module level i.e. one `__author__` for all the estimators in a given file. It might be a good idea for the `sktime` team to discuss this internally and agree on a strategy and format to use for author info across the entire library.

Anyway, for now I suggest you might use something like the below to handle the different cases and get some nicely written author info in your table:

```python
from sktime.utils import all_estimators
from importlib import import_module
import csv

def process_author_info(author_info):
    '''turn a list of author names into a string. 
       Multiple author names will be separated by a comma, 
       with the final name always preceded by "&"'''
    if isinstance(author_info, list):
        if len(author_info) > 1:
            return ', '.join(author_info[:-1]) + ' & ' + author_info[-1]
        else:
            return author_info[0]
    else:
        return author_info

with open("author_list.csv", "w+", newline='') as csvfile:
     write = csv.writer(csvfile, delimiter=",")
     write.writerow(["Class_Name","Class_ID", "Authors"])
     for modname, modclass in all_estimators():
         try:
             write = csv.writer(csvfile, delimiter = ",")
             write.writerow([modname,modclass, 
                             process_author_info(modclass.__author__)]) # <--- apply the function here
         except AttributeError:
             try:
                 write = csv.writer(csvfile, delimiter = ",")
                 write.writerow([modname,modclass, 
                                 process_author_info(import_module(modclass.__module__).__author__)]) # <--- and here

             except AttributeError:
                 write = csv.writer(csvfile, delimiter = ",")
                 write.writerow([modname,modclass, "no author info"])
```

For the above example output, this gives:

```
| Class_Name                                | Class_ID                                                                                             | Authors                                                                               |
| ----------------------------------------- | ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- |
| ARIMA                                     | <class 'sktime.forecasting.arima.ARIMA'>                                                             | Markus Löning & Hongyi Yang                                                           |
| AutoARIMA                                 | <class 'sktime.forecasting.arima.AutoARIMA'>                                                         | Markus Löning & Hongyi Yang                                                           |
| AutoCorrelationTransformer                | <class 'sktime.transformations.series.acf.AutoCorrelationTransformer'>                               | Afzal Ansari                                                                          |
| AutoETS                                   | <class 'sktime.forecasting.ets.AutoETS'>                                                             | Hongyi Yang                                                                           |
| BATS                                      | <class 'sktime.forecasting.bats.BATS'>                                                               | Martin Walter                                                                         |
| BOSSEnsemble                              | <class 'sktime.classification.dictionary_based._boss.BOSSEnsemble'>                                  | Matthew Middlehurst                                                                   |

```

**Comment by mloning** [comment]
This looks great! We can also enforce consistent formatting for `__author__` fields if that would make things simpler. Let me know what the next steps are and how we can help!

**Comment by tobyhodges** [comment]
Thanks for the feedback @mloning. Your perspective will be very helpful here: how common do you think it is/would be for a module to contain functions written by different authors? If it is a frequent occurance, or you predict that it may become one, I would suggest individual `__author__` fields for all functions in the library. If not, then the current approach is probably fine; it doesn't take much time/many extra lines of code to check and process the two cases as above.

However, I do think some standardisation of the values given for single and multiple `__authors__` would be helpful. A list is clearly appropriate for multiple authors, but what about single authors? Better to make these strings or lists of length 1? I am not intimately familiar with the project, so defer the decision to anyone who is, but my inclination is be for all `__author__` attributes to carry list values: I expect this would make it clear that it is acceptable to attribute modules and functions to multiple authors.

**Comment by fkiraly** [comment]
really nice, @afzal442!

Indeed, we may like to standardize the author field.

Replying to @tobyhodges:

>Your perspective will be very helpful here: how common do you think it is/would be for a module to contain functions written by different authors?

In the ownership model we´ve worked out, I think we assume:
* ownership (as opposed to contributorship) should extend over entire files/folders. That would correspond to the `CODEOWNERS` formalism. I think that is currently not retrieved by the approach, but would be great.
* authorship or contributorship is on file level. A file can have multiple contributors/authors.
* owners must approve and review changes to the method; if they become inactive, ownership can be transferred.

**Comment by fkiraly** [comment]
More generally, it would be nice if:
* code owners would also be retrieved
* the (scientific) type of the algorithm would be retrieved or inferred

**Comment by tobyhodges** [comment]
This is very helpful thanks @fkiraly. I was unaware of `CODEOWNERS` until you mentioned it here. I will read into it and raise for further discussion with @afzal442  & @abdulelahsm 

To make sure I am understanding correctly: by

> the (scientific) type of the algorithm would be retrieved or inferred

you mean whether the algorithm is for _classification_, _forecasting_, etc?

**Comment by fkiraly** [comment]
>you mean whether the algorithm is for classification, forecasting, etc?

precisely! And, if it's a transformer, what kind of transformer, etc.

**Comment by mloning** [comment]
@tobyhodges let's enforce `__author__` to always be of list type. We can fix the existing values in a separate PR. 

I agree with @fkiraly that it's easiest for now to assume authorship/ownership on a file level. Most major algorithms live in separate files.

**Comment by afzal442** [comment]
Thanks @fkiraly, @mloning for your feedback. Could you look around this comment that how the result looks like at https://github.com/afzal442/DocumentationEnhancementAtsktime/pull/6#issuecomment-843845485 where Toby shared his screenshot of a table of Algorithm Status in index.rst file?

**Comment by mloning** [comment]
@afzal442 that looks really good, more than happy to add this to the docs! How do you generate the underlying CSV file?

**Comment by afzal442** [comment]
Hi @mloning, @fkiraly! Now the updated code you can find as below:

```python
from sktime.utils import all_estimators
from importlib import import_module
import csv


def process_author_info(author_info):
    '''turn a list of author names into a string. 
       Multiple author names will be separated by a comma, 
       with the final name always preceded by "&"'''
    if isinstance(author_info, list):
        if len(author_info) > 1:
            return ', '.join(author_info[:-1]) + ' & ' + author_info[-1]
        else:
            return author_info[0]
    else:
        return author_info

with open("author_list.csv", "w+", newline='') as csvfile:
    write = csv.writer(csvfile, delimiter=",")
    write.writerow(["Class_Name","Class_ID","Estimator_Types", "Authors"])
    for modname, modclass in all_estimators():
        algorithm_type = ':'.join(str(modclass).split('.')[1:-2])
        try:
            author_info = process_author_info(modclass.__author__)
        except AttributeError:
            try:
                author_info = process_author_info(import_module(modclass.__module__).__author__)
            except AttributeError:
                author_info = "no author info"
        write.writerow([modname,modclass,algorithm_type,author_info])
```

Output as what it produces:

Class_Name | Class_ID | Estimator_Types | Authors
-- | -- | -- | --
ARIMA | &lt;class ‘sktime.forecasting.arima.ARIMA’&gt; | forecasting | Markus Löning &amp; Hongyi Yang
AutoARIMA | &lt;class ‘sktime.forecasting.arima.AutoARIMA’&gt; | forecasting | Markus Löning &amp; Hongyi Yang
AutoCorrelationTransformer | &lt;class ‘sktime.transformations.series.acf.AutoCorrelationTransformer’&gt; | transformations::series | Afzal Ansari
AutoETS | &lt;class ‘sktime.forecasting.ets.AutoETS’&gt; | forecasting | Hongyi Yang
BATS | &lt;class ‘sktime.forecasting.bats.BATS’&gt; | forecasting | Martin Walter
BOSSEnsemble | &lt;class ‘sktime.classification.dictionary_based._boss.BOSSEnsemble’&gt; | classification::dictionary_based | Matthew Middlehurst
BoxCoxTransformer | &lt;class ‘sktime.transformations.series.boxcox.BoxCoxTransformer’&gt; | transformations::series | Markus Löning
Catch22 | &lt;class ‘sktime.transformations.panel.catch22_features.Catch22’&gt; | transformations::panel | Matthew Middlehurst
ColumnConcatenator | &lt;class ‘sktime.transformations.panel.compose.ColumnConcatenator’&gt; | transformations::panel | Markus Löning &amp; Sajay Ganesh
ColumnEnsembleClassifier | &lt;class ‘sktime.classification.compose._column_ensemble.ColumnEnsembleClassifier’&gt; | classification::compose | Aaron Bostrom



Few snapshots of output if we let this produce through index.rst for now:

![c2](https://user-images.githubusercontent.com/11625672/119229287-e2a3a100-bb34-11eb-930f-3e0ff01a64b9.JPG)
![c3](https://user-images.githubusercontent.com/11625672/119229325-18e12080-bb35-11eb-8402-09fc48a36d66.JPG)

**Comment by afzal442** [comment]
@mloning for code-owners to be included and achievable, could you remove .py linked with some of the class strings and also remove a space among some lines of the code? like https://github.com/alan-turing-institute/sktime/blob/8f15cc3b400ec5d557ddd6de95ae9aba89770168/CODEOWNERS#L17

```md
sktime/classification/dictionary_based/tests @patrickzib

sktime/transformers/panel/dictionary_based/ @patrickzib
```

**Comment by fkiraly** [comment]
Brilliant, @afzal442 & @abdulelahsm!

I was wondering, is there any way to let users order or filter the table, like excel does?

Probably rst can't do this, but could the table be externally linked by the main docs, say?

**Comment by afzal442** [comment]
Thanks @fkiraly for your remarks.

> is there any way to let users order or filter the table, like excel does?

I think using sphinx, this is not achievable.

> could the table be externally linked by the main docs, say?

Yes Sir, this is possible. Should we link the class name or class id to its docs file?

**Comment by afzal442** [comment]
Thanks @tobyhodges for your suggestion during the discussion. Plz have a look at the screenshot below after rendering CSV to HTML page using [csvtotable](https://github.com/vividvilla/csvtotable/) tool.

![c2](https://user-images.githubusercontent.com/11625672/119538918-361e2500-bda9-11eb-9a6a-9c588273f968.JPG)

**Comment by fkiraly** [comment]
@afzal442, @abdulelahsm, @tobyhodges, I'm impressed!

Now all that remains is to plug things together - I guess we can use GitHub actions for that?

**Comment by afzal442** [comment]
Yes @fkiraly! We have been able to develop a Web page for estimator overview. And now our steps would be
![c4](https://user-images.githubusercontent.com/11625672/119684775-670e6080-be62-11eb-802c-17ceb7108220.JPG)

- [ ] how we can link each class name to its code if possible. Cuz that would be easier for a user to directly go to the respective code.
- [ ] how we can use gh actions for that

**Comment by mloning** [comment]
Great work @afzal442, @abdulelahsm and @tobyhodges! Looking forward to seeing it in action. Let me know if you any questions about setting up GitHub Actions.

**Comment by fkiraly** [comment]
Indeed, looks brilliant!
I wonder where the best place for a link is in the docs.

**Comment by afzal442** [comment]
Thanks again Markus, Franz. A huge shout out to Toby for building our foundation, helping us reach our milestone.

> I wonder where the best place for a link is in the docs.

I think we can create another file namely estimator_overview.rst where we can link our HTML file.

**Comment by afzal442** [comment]
Latest update to our work, plz have a look at this:


https://user-images.githubusercontent.com/11625672/120536869-3ac68700-c402-11eb-9587-97e4b6dca942.mp4


We need GH action to do the rest.

**Comment by fkiraly** [comment]
@afzal442, @abdulelahsm and @tobyhodges, is there a pull request for this? Might be a good time to get some reviews.

**Comment by afzal442** [comment]
Sorry @fkiraly! Let me draft a PR for this now.

**Comment by mloning** [comment]
Given the issues in #930, would it perhaps be better to merge a static version of the table first? That may be easier to implement following some other examples of creating tables for the docs inside `conf.py` (see e.g. https://github.com/scikit-learn/scikit-learn/blob/5f32f8b846382a595aa97f4227e8673aa640840e/doc/conf.py#L467).

**Comment by mloning** [comment]
Fixed by #1138 

@tobyhodges Quick update: we managed to include a searchable table which is updated automatically using sphinx and readthedocs directly. In brief, you can run hook into the sphinx build and run custom functions which we do to generate the curve and then add custom javascript to the extra files that sphinx uses to render the HTML. For details, see #1138. Thanks again for your help!

**Comment by tobyhodges** [comment]
This looks so great! Excellent work folks 🚀 It was a pleasure to be (briefly) involved in this project 🙌
